KOGITO-513 - Do not open editor when .bpmn file can't be loaded

hasys opened the pull request:
Hi @romartin, @domhanak,

To test this PR just put `&` sign inside of any XML attribute value. Only Kogito code was modified so no need to test Business Central. 

![Screenshot from 2020-06-16 14-23-15](https://user-images.githubusercontent.com/1477262/84776325-76452700-afe0-11ea-910f-997dc93754b2.png)
![Screenshot from 2020-06-16 14-25-27](https://user-images.githubusercontent.com/1477262/84776332-780eea80-afe0-11ea-9bc0-f096416ac9dd.png)


VSIX can be downloaded here: https://drive.google.com/drive/u/0/folders/1UzvsBmC4PVaxaIFdmnetQjwRizJJuOjl

PR is ready for review, thank you!

<details>
  <summary>Technical details</summary>
  <pre>There are two parts fixed: BPMN Editor itself and HTML where editor is embedded. 
* Kogito-Stunner has generic AbstractEditor which already support ACE Editor but only for `DiagramParsingException`, so I fixed it in BPMN Editor and ACE editor was called.
* Second issue was that we need to link ACE editor in HTML separately from the editor itself, that's why there is HTML fix as well.

Also, looks like there are two different ways how ACE Editor injected to the HTML. For Kogito Showcase it is `index.html` from `kie-wb-common-stunner-bpmn-kogito-runtime` module but for VS Code/Github ACE editor configureed here: https://github.com/kiegroup/kogito-tooling/blob/master/packages/kie-bc-editors/src/GwtEditorRoutes.ts#L58

I fixed showcase one, and it was already good on VS Code plugin side. </pre>
</details>

hasys commented:
Jenkins execute full downstream build

manstis commented:
@romartin @porcelli @ederign @caponetto @karreiro 

> Approving as it will allow users to manage "broken" processes, at least, although as commented I think we've to work now on completely dropping the use of the ACE editor and just relying on the Monaco one, at least for the BPMN editor.

Please include DMN with any changes re: ACE (as they should be consistent). I suspect changes would be largely at the Stunner level and hence DMN, CM and BPMN would all benefit. Furthermore DMN already has Monaco embedded (as an expression editor in DMN's table cells). IIRC it's footprint was not small either. 

It's probably worth investigation to see if we can re-use the Monaco component in VSCode to fallback to if a file cannot be opened. This would require the editors signalling to the envelope that there was a problem and for an alternative editor to be opened. The same mechanism would also work for other channels but it may require the envelope to ask the editor to re-open the file (with a different component). Anyway... idle musings for consideration.

hasys commented:
Hi @domhanak,


test fails in Full downstream build doesn't related to this PR:

<details>
  <summary>Failed tests</summary>
  <pre>

org.kie.processmigration.it.ProcessMigrationIntegrationTest.testReadinessCheck
--------------------------------------------------------------------------------------------------------------------------------------

Error Message

Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)

Stacktrace

org.apache.http.conn.HttpHostConnectException: Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testReadinessCheck(ProcessMigrationIntegrationTest.java:114)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testReadinessCheck(ProcessMigrationIntegrationTest.java:114)

org.kie.processmigration.it.ProcessMigrationIntegrationTest.testMigration
--------------------------------------------------------------------------------------------------------------------------------------

Error Message

Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)

Stacktrace

org.apache.http.conn.HttpHostConnectException: Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.createPlan(ProcessMigrationIntegrationTest.java:179)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.createMigration(ProcessMigrationIntegrationTest.java:148)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testMigration(ProcessMigrationIntegrationTest.java:134)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.createPlan(ProcessMigrationIntegrationTest.java:179)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.createMigration(ProcessMigrationIntegrationTest.java:148)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testMigration(ProcessMigrationIntegrationTest.java:134)

org.kie.processmigration.it.ProcessMigrationIntegrationTest.testLivenessCheck
-------------------------------------------------------------------------------------------------------------------------------------

Error Message

Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)

Stacktrace

org.apache.http.conn.HttpHostConnectException: Connect to localhost:8180 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testLivenessCheck(ProcessMigrationIntegrationTest.java:108)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at org.kie.processmigration.it.ProcessMigrationIntegrationTest.testLivenessCheck(ProcessMigrationIntegrationTest.java:108)

</pre>
</details>

PR is ready for review. Thank you!

domhanak commented:
@hasys using the VSIX file I am unable to load any CORRECT BPMN file, the editor just keeps on loading. When I load broken file, the Text editor is shown as expected.

hasys commented:
Hi @domhanak, it was caused by outdated version of VSIX, Visual Studio updated and VSIX became outdated, I rebased my changes to the master and it started to work.

BUT I found another issue, that it was not possible to save changes in XML editor, so I fixed it as well.
See: https://issues.redhat.com/secure/attachment/12475804/xmlEditor.webm

You can download the latest VSIX here: https://drive.google.com/drive/u/0/folders/1UzvsBmC4PVaxaIFdmnetQjwRizJJuOjl

Or you can rebase/recompile if it will be outdated again. Thank you!

domhanak commented:
Ahh, nice.  Works well now, I was able to fix the BPMN, save it and reopen within our editor

domhanak commented:
jenkins retest this

domhanak commented:
jenkins retest this

domhanak commented:
jenkins retest this

hasys commented:
Hi @romartin, PR is ready to be merged. Thank you, @romartin, @domhanak!
